### PR TITLE
Switch from rust-websocket to ws (fixes #14517)

### LINKED ISSUE #14517: Migrate from rust-websocket to ws-rs
rust-websocket appears to not be maintained anymore. Possibly migrate to ws-rs? This is sorta blocking the rust-openssl bump: https://github.com/servo/servo/pull/14484

https://github.com/cyderize/rust-websocket

https://github.com/housleyjk/ws-rs

### PR DESCRIPTION
**Pull request by nox**
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/16012)

### CODE REVIEW

**Comment by nox** [comment]
```
error[E0432]: unresolved import `libc::EPOLLRDHUP`
 --> /home/servo/.cargo/registry/src/github.com-1ecc6299db9ec823/mio-0.6.4/src/sys/unix/epoll.rs:8:32
  |
8 | use libc::{EPOLLERR, EPOLLHUP, EPOLLRDHUP, EPOLLONESHOT};
  |                                ^^^^^^^^^^ no `EPOLLRDHUP` in the root. Did you mean to use `EPOLLHUP`?

error[E0432]: unresolved import `libc::EPOLLONESHOT`
 --> /home/servo/.cargo/registry/src/github.com-1ecc6299db9ec823/mio-0.6.4/src/sys/unix/epoll.rs:8:44
  |
8 | use libc::{EPOLLERR, EPOLLHUP, EPOLLRDHUP, EPOLLONESHOT};
  |                                            ^^^^^^^^^^^^ no `EPOLLONESHOT` in the root
```

**Comment by nox** [comment]
Seems like this is fixed in mio 0.6.5, but I'll check the other job results before retrying.

**Comment by nox** [comment]
I'm also really confused as to why this didn't fail in the past.

**Comment by nox** [comment]
https://github.com/housleyjk/ws-rs/issues/103

**Comment by staabm** [comment]
nit: this PR changes from `servo-websocket` to `ws` not `rust-websocket` as the PR-title mentions

**Comment by jdm** [comment]
servo-websocket is just a fork that we made last week.

**Comment by nox** [comment]
This switch is very crude btw, and doesn't benefit from how ws-rs is able to handle multiple websocket connections over a single thread. Given we are just trying to unblock that hyper/openssl bump, I don't think we should block here.

**Comment by nox** [comment]
https://github.com/housleyjk/ws-rs/pull/119

**Comment by jdm** [comment]
```
  ▶ Unexpected subtest result in /websockets/interfaces/WebSocket/close/close-connecting.html:
  └ PASS [expected FAIL] WebSockets: close() when connecting
```

**Comment by nox** [comment]
@jdm Unfortunately this test does not pass in Firefox and Chrome, should I mark it as PASS and file an issue on wpt?

**Comment by nox** [comment]
Also:

```
  ▶ Unexpected subtest result in /websockets/cookies/007.html:
  │ FAIL [expected PASS] WebSockets: when to process set-cookie fields in ws response
  │   → assert_unreached: cookie was set during script execution Reached unreachable code
  │ 
  │ @http://web-platform.test:8000/websockets/cookies/007.html:31:5
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1409:20
  │ async_test@http://web-platform.test:8000/resources/testharness.js:518:13
  └ @http://web-platform.test:8000/websockets/cookies/007.html:12:1
```

**Comment by jdm** [comment]
Filing an issue on WPT sounds good to me.

File: components/net/websocket_loader.rs
Comment (jdm) [inline components/net/websocket_loader.rs]: This changes the current behaviour and goes against the spec - cookies are set from the queued task, rather than immediately on connection: https://dxr.mozilla.org/servo/rev/e8b364793416465f6b4877414623dac2c52f2e6a/components/script/dom/websocket.rs#500-505

File: components/net/websocket_loader.rs
Comment (jdm) [inline components/net/websocket_loader.rs]: Either the first condition should be inverted the && should be ||.

File: components/net/websocket_loader.rs
Comment (jdm) [inline components/net/websocket_loader.rs]: Why don't we need this any more? It was added in 27b4ac61508877e6ded6f01a65bd02e01dc4e81b to address the case where we should not echo the server's close message if we initiated the closure.

File: components/net/websocket_loader.rs
Comment (jdm) [inline components/net/websocket_loader.rs]: We should be echoing the message back to the server here in [some cases](https://tools.ietf.org/html/rfc6455#section-1.4); https://docs.rs/ws/0.6.0/ws/trait.Handler.html#method.on_close doesn't show this being handled automatically.

**Review by jdm — CHANGES_REQUESTED** [review]
Generally good, but it appears that we are regressing some cases that we addressed in the old code.

**Review by nox — COMMENTED** [review]
Addressed your comments.

**Comment by nox** [comment]
This is still blocked on https://github.com/housleyjk/ws-rs/pull/120.

File: components/net/websocket_loader.rs
Comment (jdm) [inline components/net/websocket_loader.rs]: Got a link demonstrating this?

**Comment by nox** [comment]
I'm going to switch to tungstenite instead.

**Pull request closed without merging**
